marklogic-client-api/src/main/java/com/marklogic/client/impl/Utilities.java

```diff
@@ -669,4 +675,26 @@ static public DatatypeFactory getDatatypeFactory() {
     }
     return datatypeFactory;
   }
+  
+  static public void write(InputStream in, Path path) { 
+      try {
+      RandomAccessFile outFile = new RandomAccessFile(path.toString(), "rw");
+      FileChannel  inChannel = outFile.getChannel();
+      byte[] buf = new byte[BUFFER_SIZE * 2];
+      ByteBuffer byteBuf;
+      
+      while (in.read(buf) != -1) {
+        byteBuf = ByteBuffer.wrap(buf);
+        byteBuf.flip();
+        inChannel.write(byteBuf);
+        byteBuf.clear();
+      }
+      outFile.close();
+      } catch (IOException e) {
+          throw new MarkLogicIOException("Internal Exception occured.");
+      } 
```

Comment:
Might want to add a 'finally { outfile.close() } ' if it exists to just limit the time it takes to close out the file handle.

Also, a large initial direct byte buffer will lessen the impact buffer allocation has if one direct buffer is created for universal use, then 'slices' are taken off to write files then returning the memory when the write occurs in equal chunks. This will reuse native memory outside the java heap at a lower level and should give a noticeable boost if writing binaries that are significantly above the L1+L2 cache size (which binaries tend to be). 

Just a few random thoughts.

Reply:
Thanks Alex.!
So you mean we should have something like - 
byte[] buf = new byte[in.available()];
ByteBuffer byteBuf = ByteBuffer.allocate(in.available());
if(in.read(buf) != -1) {
      byteBuf = ByteBuffer.wrap(buf);
 }
and then write out the byteBuffer in chunks using FileChannel. 
Am i correct.?

Reply:
Nah, meaning DirectByteBuffer (ByteBuffer.allocateDirect()) and reuse it as your bytebuffer throughout (allocate once, reuse multiple). A direct allocation should be significantly faster than a non-direct buffer writing to channels, as the native OS level code doesn't have to keep dipping back into the JVM to read, instead it can pull from the direct buffer. 

Though if we're wanting to go pure performance, @ehennum, do we want to consider off heap for the reads/writes from disk? It's in unsafe space, but so is technically a direct allocation of memory, and would greatly improve the read performance (as well as the write)?